README.md
!["PREVIEW"](preview.jpg)

[Sword of PastJoys](https://steamcommunity.com/sharedfiles/filedetails/?id=2571573233), created based on DST mod “[辞欢](https://steamcommunity.com/sharedfiles/filedetails/?id=2473285695)”

This mod adds abundant configurations that player can manually set, and meanwhile it decreases the difficulty for level up of the sword.


changelog.txt
-- 20210814 0.02c
update sword-hold texture

-- 20210814 0.02b
bug fixs

-- 20210811 0.02a
bug fixed

-- 20210811 0.02
1. add English supported
2. add AOE effect

-- 20210811 0.01
"Sword of PastJoys"


modmain.lua
--GLOBAL相关照抄
GLOBAL.setmetatable(
  env,
  {
    __index = function(t, k)
      return GLOBAL.rawget(GLOBAL, k)
    end
  }
)

local config = {
  init_atk = GetModConfigData("init_atk"),
  atk_range = GetModConfigData("atk_range"),
  max_level = GetModConfigData("max_level"),
  lvlup_ways = GetModConfigData("lvlup_ways"),
  level_atk_growth = GetModConfigData("level_atk_growth"),
  addtional_atk = GetModConfigData("addtional_atk"),
  aoe = GetModConfigData("aoe"),
  health_steal = GetModConfigData("health_steal"),
  sanity_steal = GetModConfigData("sanity_steal"),
  waterproofer = GetModConfigData("waterproofer"),
  insulator = GetModConfigData("insulator"),
  light = GetModConfigData("light"),
  blink = GetModConfigData("blink"),
  recipe = GetModConfigData("recipe"),
  lang = GetModConfigData("language")
}
GLOBAL.config = config

local json = GLOBAL.require "libs.json"
local helpers = GLOBAL.require "libs.helpers"

print("user:config:", json.encode(GLOBAL.config))
GLOBAL.lang = helpers.init_language()

PrefabFiles = {
  "wuqi"
}

Assets = {}

STRINGS.NAMES.WUQI = GLOBAL.lang.weapon_name
STRINGS.RECIPE_DESC.WUQI = GLOBAL.lang.weapon_recipe_desc
STRINGS.CHARACTERS.GENERIC.DESCRIBE.WUQI = GLOBAL.lang.characters_generic_desc

--武器
local recipe = {}
if GLOBAL.config.recipe == "easy" then
  recipe = {Ingredient("axe", 1), Ingredient("axe", 1), Ingredient("axe", 1)}
else
  recipe = {Ingredient("nightsword", 2), Ingredient("nightstick", 3), Ingredient("trident", 1)}
end

AddRecipe(
  "wuqi", --添加物品的配方
  recipe,
  RECIPETABS.WAR,
  TECH.SCIENCE_TWO, --制作栏和解锁的科技（这里是战斗，需要科学二本）
  nil,
  nil,
  nil,
  nil,
  nil, --是否有placer  是否有放置的间隔  科技锁  制作的数量（改成2就可以一次做两个） 需要的标签（比如女武神的配方需要女武神的自有标签才可以看得到）
  "images/inventoryimages/wuqi.xml", --配方的贴图（跟物品栏使用同一个贴图）
  "wuqi.tex"
)

STRINGS.RECIPE_DESC.GNARWAIL_HORN = GLOBAL.lang.gnarwail_horn_recipe_desc
AddRecipe(
  "gnarwail_horn",
  {Ingredient("horn", 1), Ingredient("lightninggoathorn", 1), Ingredient("bathbomb", 5)},
  RECIPETABS.REFINE,
  TECH.SCIENCE_TWO
)


scripts/libs/helpers.lua
local json = require "libs.json"
local helpers = {}

-- @return `Table` object
helpers.get_keys = function(_table)
    local keyset = {}
    local n = 0
    for k, v in pairs(_table) do
        n = n + 1
        keyset[n] = k
    end
    return keyset
end

local en = {
    weapon_name = "Sword of PastJoys",
    weapon_recipe_desc = "Past joyes going throght the late night",
    characters_generic_desc = "Can't sank on those memories right now",
    gnarwail_horn_recipe_desc = "Protect animals, make world better.",
    weapon_gameon_name = "Sword of PastJoys(Lv. %s)",
    weapon_level_up = "Sword of PastJoys is level up by %s"
}

local zh = {
    weapon_name = "辞欢",
    weapon_recipe_desc = "惊鸿难抵梦重重",
    characters_generic_desc = "辞去今朝别往事",
    gnarwail_horn_recipe_desc = "保护珍惜动物，从我做起",
    weapon_gameon_name = "【辞欢剑】当前等级：%s",
    weapon_level_up = "辞欢等级加%s"
}

helpers.init_language = function()
    local language = config.lang or "Chinese"
    if language == "Chinese" then
        return zh
    else
        return en
    end
end

return helpers


scripts/prefabs/wuqi.lua
local attack = require "prefabs.attack"

local assets = {
    Asset("ANIM", "anim/wuqi.zip"), --地上的动画
    Asset("ANIM", "anim/swap_wuqi.zip"), --手里的动画
    Asset("ATLAS", "images/inventoryimages/wuqi.xml"), --加载物品栏贴图
    Asset("IMAGE", "images/inventoryimages/wuqi.tex")
}

local on_season_changed = function(insulator)
    if TheWorld.state.temperature > 35 then
        insulator:SetSummer()
    else
        insulator:SetWinter()
    end
end

local function onequip(inst, owner) --装备
    owner.AnimState:OverrideSymbol("swap_object", "swap_wuqi", "swap_wuqi")
    --替换的动画部件	使用的动画	替换的文件夹（注意这里也是文件夹的名字）
    owner.AnimState:Show("ARM_carry")
    owner.AnimState:Hide("ARM_normal")
    -- Splash attack(AOE)
    if config.aoe ~= 0 then
        owner.components.combat:SetAreaDamage(3, config.aoe)
    end
    if config.light ~= 0 then
        inst.Light:Enable(true)
    end
end

local function onunequip(inst, owner) --解除装备
    owner.AnimState:Hide("ARM_carry")
    owner.AnimState:Show("ARM_normal")
    if config.aoe ~= 0 then
        owner.components.combat:SetAreaDamage(-3, -config.aoe)
    end
end

local function onsave(inst, data)
    data.level = inst.level or 0
    if config.insulator then
        on_season_changed(inst.components.insulator)
    end
end

local function onpreload(inst, data)
    if not data or not data.level then
        return
    end
    attack.update_weapon(inst, data.level)
end

local function fn()
    local inst = CreateEntity()

    inst.entity:AddTransform()
    inst.entity:AddAnimState()
    inst.entity:AddNetwork()
    if config.light ~= 0 then
        inst.entity:AddLight() -- add light
    end

    MakeInventoryPhysics(inst)

    -- set light (unequipment state)
    if config.light ~= 0 then
        inst.Light:SetIntensity(0.5)
        inst.Light:SetRadius(config.light or 5)
        inst.Light:Enable(true)
        inst.Light:SetFalloff(1)
        inst.Light:SetColour(200 / 255, 200 / 255, 200 / 255)
    end

    inst.AnimState:SetBank("wuqi")
    inst.AnimState:SetBuild("wuqi")
    inst.AnimState:PlayAnimation("idle")

    inst:AddTag("sharp") --武器的标签跟攻击方式跟攻击音效有关 没有特殊的话就用这两个
    inst:AddTag("pointy")

    inst:AddComponent("talker")
    inst.components.talker.fontsize = 40
    inst.components.talker.font = NUMBERFONT
    inst.components.talker.colour = Vector3(.9, .4, .4)
    inst.components.talker.offset = Vector3(0, 0, 0)
    inst.components.talker.symbol = "swap_object"

    MakeInventoryFloatable(inst, "med", 0.05, {1.1, 0.5, 1.1}, true, -9)

    inst.entity:SetPristine()

    if not TheWorld.ismastersim then
        return inst
    end

    inst.level = 0

    inst:AddComponent("weapon") --增加武器组件 有了这个才可以打人
    inst.components.weapon:SetDamage(config.init_atk) --设置伤害
    inst.components.weapon:SetOnAttack(attack.on_atk)

    inst:AddComponent("inspectable") --可检查组件

    inst:AddComponent("inventoryitem") --物品组件
    inst.components.inventoryitem.atlasname = "images/inventoryimages/wuqi.xml" --物品贴图

    inst:AddComponent("equippable") --可装备组件
    inst.components.equippable:SetOnEquip(onequip)
    inst.components.equippable:SetOnUnequip(onunequip)
    inst.components.equippable.walkspeedmult = 1.25 -- 给装备附加额外的移速

    inst:AddComponent("named")
    inst.components.named:SetName(string.format(lang.weapon_gameon_name, 0))

    -- 攻击距离
    inst.components.weapon:SetRange(config.atk_range, config.atk_range)

    MakeHauntableLaunch(inst)

    inst.OnSave = onsave
    inst.OnPreLoad = onpreload

    -- 设定隔热
    if config.insulator then
        inst:AddComponent("insulator")
        on_season_changed(inst.components.insulator)
        inst.components.insulator:SetInsulation(280)
    -- inst:ListenForEvent("seasonChange", on_season_changed(inst.components.insulator), TheWorld)
    end

    -- 防水
    if config.waterproofer then
        inst:AddComponent("waterproofer")
        inst.components.waterproofer:SetEffectiveness(1)
    end

    -- Blink
    if config.blink then
        inst:AddComponent("blinkstaff")
    end

    return inst
end

return Prefab("wuqi", fn, assets)


scripts/prefabs/attack.lua
local json = require "libs.json"
local helpers = require "libs.helpers"
local attak = {}

local init_atk = config.init_atk or 17
local level_atk_growth = config.level_atk_growth or 1
local max_level = config.max_level or 300
local lvlup_ways = config.lvlup_ways or "moderate"
local addtional_atk = config.addtional_atk or "0" -- 'current' | 'max' | '0'
local life_steal_rate = config.health_steal or 0
local sanity_steal_rate = config.sanity_steal or 0

-- @dmg: number
-- @player: `Player` object
function __deal_life_steal(dmg, player)
    dmg = dmg or 0
    if not player or dmg == 0 then
        return
    end

    if life_steal_rate ~= 0 then
        player.components.health:DoDelta(life_steal_rate * dmg)
    end

    if sanity_steal_rate ~= 0 then
        player.components.sanity:DoDelta(sanity_steal_rate * dmg)
    end
end

-- 经典升级模型
-- @return lvlup: number
function __classical_lvlup(target_health, randint)
    local lvlup = 0
    if target_health >= 20000 then
        --对于血量大于等于20000的怪物来说，辞欢 20%概率 +1等级
        if randint <= 20 then
            lvlup = 1
        end
    elseif target_health >= 2000 then
        --对于血量大于等于2000，小于20000, 10%概率 +2级
        if randint <= 10 then
            lvlup = 2
        end
    elseif target_health >= 200 then
        --对于血量大于等于200，小于2000, 5%概率 +1级
        if randint <= 5 then
            lvlup = 1
        end
    else
        --对于血量小于200的怪物来说，辞欢1%概率 +1级
        if randint == 1 then
            lvlup = 1
        end
    end

    return lvlup
end

-- update Sword of PastJoys damage
function __do_weapon_update(inst, target_level)
    inst.level = target_level
    local dmg = init_atk + level_atk_growth * target_level
    inst.components.weapon:SetDamage(dmg)
    inst.damage = dmg
    inst.components.named:SetName(string.format(lang.weapon_gameon_name, target_level))
end

-- @target: the Monster we killed
function __after_target_dead(inst, owner, target)
    -- print("if target dead YET?", target.components.health:IsDead())
    if not target.components.health:IsDead() then
        return
    end

    local current_level = inst.level
    if max_level ~= 0 and current_level >= max_level then
        inst.level = max_level
        return __do_weapon_update(inst, max_level)
    end

    local target_health = target.components.health.maxhealth
    local randint = math.random(1, 100)
    local lvlup = 0
    -- print("target health:", target_health)
    -- print("randint:", randint)
    if lvlup_ways == "easy" and randint <= 65 then
        lvlup = 1
    elseif lvlup_ways == "moderate" and target_health >= 200 and randint <= 33 then
        lvlup = 1
    elseif lvlup_ways == "classic" then
        lvlup = __classical_lvlup(target_health, randint)
    end

    if lvlup ~= 0 then
        __do_weapon_update(inst, inst.level + lvlup)
        owner.SoundEmitter:PlaySound("dontstarve/wilson/equip_item_gold")
        inst.components.talker:Say(string.format(lang.weapon_level_up, lvlup))
    end
end

function __calculate_additional_dmg(rate, target)
    if addtional_atk == "current" then
        return target.components.health.currenthealth * rate
    elseif addtional_atk == "max" then
        return target.components.health.maxhealth * rate
    else
        return 0
    end
end

function on_atk(inst, owner, target)
    if not inst.level then
        __do_weapon_update(inst, 0)
    elseif not inst.damage then
        __do_weapon_update(inst, inst.level)
    end

    -- calculate damage
    local rate = 0
    if inst.level >= 60 then
        rate = 0.02
    elseif inst.level >= 30 then
        rate = 0.01
    end

    local addtional_dmg = __calculate_additional_dmg(rate, target)
    if addtional_dmg ~= 0 then
        -- print("造成额外伤害:", addtional_dmg)
        target.components.health:DoDelta(-addtional_dmg, nil, nil, true, nil, true)
    end

    local dmg = inst.damage
    dmg = dmg + addtional_dmg

    __deal_life_steal(dmg, owner)
    __after_target_dead(inst, owner, target)
end

attak.on_atk = on_atk
attak.update_weapon = __do_weapon_update
return attak


modinfo.lua
author = "EvinK"
version = "0.0.2c"
name = "辞欢+"
description = [["辞欢+"基于DST模组【辞欢】而创作，本模组提供了大量可自定义内容，同时降低辞欢剑升级难度

希望你喜欢 ：）

当前版本:]] .. version

forumthread = ""

dst_compatible = true --兼容联机
dont_starve_compatible = false --不兼容单机
reign_of_giants_compatible = false --不兼容巨人
all_clients_require_mod = true --所有人mod

api_version = 10

icon_atlas = "modicon.xml"
icon = "modicon.tex"

server_filter_tags = {"cihuan"} --服务器标签可以不写

local lang =
	ChooseTranslationTable(
	{
		"English",
		["zh"] = "Chinese"
	}
)

configuration_options = {}

if lang == "Chinese" then
	configuration_options = {
		{
			name = "init_atk",
			label = "初始攻击力",
			hover = "",
			options = {
				{description = "17", data = 17},
				{description = "70", data = 70}
			},
			default = 17
		},
		{
			name = "atk_range",
			label = "攻击范围",
			hover = "默认两格，标准为1格",
			options = {
				{description = "1格", data = 1},
				{description = "2格", data = 2}
			},
			default = 2
		},
		{
			name = "max_level",
			label = "武器等级",
			hover = "",
			options = {
				{description = "限制61级", data = 61},
				{description = "限制300级", data = 300},
				{description = "不限", data = 0}
			},
			default = 300
		},
		{
			name = "lvlup_ways",
			label = "升级调控",
			hover = "简单: 击杀怪物有65%的几率升级  经典：原版难度 \n适中: 击杀200血及以上的怪物有33%的几率升级",
			options = {
				{description = "简单", data = "easy"},
				{description = "适中", data = "moderate"},
				{description = "经典", data = "classic"}
			},
			default = "moderate"
		},
		{
			name = "level_atk_growth",
			label = "伤害成长",
			hover = "",
			options = {
				{description = "每级1点", data = 1},
				{description = "每级3点", data = 3}
			},
			default = 1
		},
		{
			name = "addtional_atk",
			label = "附加伤害",
			hover = "每30级按照怪物的当前/最大血量百分比附加额外伤害\n 最多叠加2%（附加伤害也会造成吸血效果）",
			options = {
				{description = "1%当前生命值", data = "current"},
				{description = "1%最大生命值", data = "max"},
				{description = "不附加伤害", data = "0"}
			},
			default = "current"
		},
		{
			name = "aoe",
			label = "溅射伤害",
			hover = "3格范围内的每次攻击造成群体伤害\n默认关闭",
			options = {
				{description = "45%", data = 0.45},
				{description = "25%", data = 0.25},
				{description = "关闭", data = 0}
			},
			default = 0
		},
		{
			name = "health_steal",
			label = "吸血",
			hover = "每次攻击回复一定量的HP\n依玩家设置而定",
			options = {
				{description = "9%", data = 0.09},
				{description = "18%", data = 0.18},
				{description = "45%", data = 0.45},
				{description = "100%(打多少回多少)", data = 1},
				{description = "关", data = 0}
			},
			default = 0.09
		},
		{
			name = "sanity_steal",
			label = "精神回复",
			hover = "每次攻击回复一定量的精神值\n依玩家设置而定",
			options = {
				{description = "9%", data = 0.09},
				{description = "18%", data = 0.18},
				{description = "45%", data = 0.45},
				{description = "100%(打多少回多少)", data = 1},
				{description = "关", data = 0}
			},
			default = 0.09
		},
		{
			name = "waterproofer",
			label = "防水",
			hover = "默认打开",
			options = {
				{description = "开", data = true},
				{description = "关", data = false}
			},
			default = true
		},
		{
			name = "insulator",
			label = "隔热",
			hover = "默认打开",
			options = {
				{description = "开", data = true},
				{description = "关", data = false}
			},
			default = true
		},
		{
			name = "light",
			label = "光照",
			hover = "在夜晚提供光照\n强光(默认)约5半径，弱光约2半径",
			options = {
				{description = "强光", data = 5},
				{description = "弱光", data = 2},
				{description = "关闭", data = 0}
			},
			default = 5
		},
		{
			name = "blink",
			label = "闪现",
			hover = "闪现功能",
			options = {
				{description = "开", data = true},
				{description = "关", data = false}
			},
			default = true
		},
		{
			name = "recipe",
			label = "制作图纸",
			hover = "",
			options = {
				{description = "简单", data = "easy"},
				{description = "经典(原版)", data = "classic"}
			},
			default = "easy"
		},
		{
			name = "language",
			label = "语言",
			hover = "根据您的语言设置，模组已被设置为中文",
			options = {
				{description = lang, data = lang}
			},
			default = lang
		}
	}
else
	name = "Sword of PastJoys"
	description =
		[["Sword of PastJoys" created based on DST mod "辞欢"

the mod adds abundant configurations that player can manually set, and meanwhile it decreases the difficulty for level up of the sword.

Hope you enjoy Sword of PastJoys :)

current version: ]] ..
		version .. [[   contact: EvinK@foxmail.com]]
	configuration_options = {
		{
			name = "init_atk",
			label = "Born Attack",
			hover = "",
			options = {
				{description = "17", data = 17},
				{description = "70", data = 70}
			},
			default = 17
		},
		{
			name = "atk_range",
			label = "Attack Range",
			hover = "Default set to 2 units \n Game standard is 1 unit.",
			options = {
				{description = "1 unit", data = 1},
				{description = "2 units", data = 2}
			},
			default = 2
		},
		{
			name = "max_level",
			label = "Sword Level Limit",
			hover = "",
			options = {
				{description = "Limit to Lv.61", data = 61},
				{description = "Limit to Lv.300", data = 300},
				{description = "Unlimit", data = 0}
			},
			default = 300
		},
		{
			name = "lvlup_ways",
			label = "Sword Level-up Mode",
			hover = "Easy: 65% chance to get a level-up  Classic: original ways (harder) \n Moderate: 33% chance to level up (target HP >= 200)",
			options = {
				{description = "Easy", data = "easy"},
				{description = "Moderate", data = "moderate"},
				{description = "Classic", data = "classic"}
			},
			default = "moderate"
		},
		{
			name = "level_atk_growth",
			label = "Damage Growth",
			hover = "Every time the sword is level up, its basic damage will increase.",
			options = {
				{description = "+1 damage", data = 1},
				{description = "+3 damage", data = 3}
			},
			default = 1
		},
		{
			name = "addtional_atk",
			label = "Extra Damage",
			hover = "Provide extra damage based on target percent HP in every 30 levles.\n Limits to 2%（it works for life/sanity stealing too)",
			options = {
				{description = "1% Current HP", data = "current"},
				{description = "1% Max HP", data = "max"},
				{description = "No Extra Damage", data = "0"}
			},
			default = "current"
		},
		{
			name = "aoe",
			label = "AOE Damage",
			hover = "A splash attack in 3 units. default set to OFF",
			options = {
				{description = "45%", data = 0.45},
				{description = "25%", data = 0.25},
				{description = "OFF", data = 0}
			},
			default = 0
		},
		{
			name = "health_steal",
			label = "Health Steal",
			hover = "Recover your HP，you can specify recover rate",
			options = {
				{description = "9%", data = 0.09},
				{description = "18%", data = 0.18},
				{description = "45%", data = 0.45},
				{description = "100%", data = 1},
				{description = "OFF", data = 0}
			},
			default = 0.09
		},
		{
			name = "sanity_steal",
			label = "Sanity Recover",
			hover = "Recover your sanity，you can specify recover rate",
			options = {
				{description = "9%", data = 0.09},
				{description = "18%", data = 0.18},
				{description = "45%", data = 0.45},
				{description = "100%", data = 1},
				{description = "OFF", data = 0}
			},
			default = 0.09
		},
		{
			name = "waterproofer",
			label = "Water Proof",
			hover = "",
			options = {
				{description = "ON", data = true},
				{description = "OFF", data = false}
			},
			default = true
		},
		{
			name = "insulator",
			label = "Insulate Temperature",
			hover = "",
			options = {
				{description = "ON", data = true},
				{description = "OFF", data = false}
			},
			default = true
		},
		{
			name = "light",
			label = "Light",
			hover = "Provide a light during night.\n Bright(default) is about 5 units，Weak is about 2 units",
			options = {
				{description = "Bright", data = 5},
				{description = "Weak", data = 2},
				{description = "OFF", data = 0}
			},
			default = 5
		},
		{
			name = "blink",
			label = "Blink",
			hover = "Allow you to jump a small distance.",
			options = {
				{description = "ON", data = true},
				{description = "OFF", data = false}
			},
			default = true
		},
		{
			name = "recipe",
			label = "Recipe",
			hover = "",
			options = {
				{description = "Simple", data = "easy"},
				{description = "Classic", data = "classic"}
			},
			default = "easy"
		},
		{
			name = "language",
			label = "Language",
			hover = "According to your Language Setting, this mod has been set to English\n Right now, there is only English/Chinese supported",
			options = {
				{description = lang, data = lang}
			},
			default = lang
		}
	}
end
